hs_core/models.py

```diff
@@ -1508,41 +1512,205 @@ class Meta:
+        # with a possible additional folder
+        # Compute FQN prefix
+        prefix = os.path.join(instance.content_object.short_id, 'data', 'contents')
+
+    if not folder:
+        folder = instance.file_folder
+
+    if folder is not None:
+        # use subfolder
+        return os.path.join(prefix, folder, filename)
     else:
-        return os.path.join(instance.content_object.short_id, 'data', 'contents', filename)
+        # use root folder
+        return os.path.join(prefix, filename)
+
+
+def _path_is_allowed(path):
+    """ paths containing '/../' are suspicious """
+    if path == "":
+        raise ValidationError("Empty file paths are not allowed")
+    if '/../' in path:
+        raise SuspiciousFileOperation("File paths cannot contain '/../'")
 
 
 class ResourceFile(models.Model):
     object_id = models.PositiveIntegerField()
     content_type = models.ForeignKey(ContentType)
-
     content_object = GenericForeignKey('content_type', 'object_id')
-    resource_file = models.FileField(upload_to=get_path, max_length=500, null=True, blank=True,
-                                     storage=IrodsStorage()
-                                     if getattr(settings, 'USE_IRODS', False)
-                                     else DefaultStorage())
+    # the file_folder attribute makes uploads work properly. It is otherwise not used.
+    file_folder = models.CharField(max_length=255, null=True)
+    resource_file = models.FileField(upload_to=get_path, max_length=500,
+                                     null=True, blank=True, storage=IrodsStorage())
     # the following optional fields are added for use by federated iRODS resources where
     # resources are created in the local federated zone rather than hydroshare zone, in
     # which case resource_file is empty, and we record iRODS logical resource file name
     # and file size for customized copy or move operations for resource creation and other
     # operations fed_resource_file in particular is a counterpart of resource_file, but
     # handles files uploaded from local disk and store the files to federated zone rather
     # than hydroshare zone.
-    fed_resource_file = models.FileField(upload_to=get_path, max_length=500, null=True, blank=True,
-                                         storage=IrodsStorage('federated') if getattr(
-                                             settings, 'USE_IRODS', False) else DefaultStorage())
+    fed_resource_file = models.FileField(upload_to=get_path, max_length=500,
+                                         null=True, blank=True, storage=IrodsStorage('federated'))
+    # DEPRECATED: utilize resfile.set_storage_path(path) and resfile.storage_path.
     fed_resource_file_name_or_path = models.CharField(max_length=255, null=True, blank=True)
+    # TODO: why is size a CharField???
     fed_resource_file_size = models.CharField(max_length=15, null=True, blank=True)
```

Comment:
@alvacouch Agreed although it works to use a string to store the file size, it makes more sense to have fed_resource_file_size a BigIntegerField rather than a CharField. Feel free to correct it as needed.